File: hub-detect/src/main/groovy/com/blackducksoftware/integration/hub/detect/workflow/project/DetectorEvaluationNameVersionDecider.java
package com.blackducksoftware.integration.hub.detect.workflow.project;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.blackducksoftware.integration.hub.detect.detector.DetectorType;
import com.blackducksoftware.integration.hub.detect.workflow.search.result.DetectorEvaluation;
import com.synopsys.integration.util.NameVersion;

public class DetectorEvaluationNameVersionDecider {
    private final Logger logger = LoggerFactory.getLogger(DetectorEvaluationNameVersionDecider.class);

    private final DetectorNameVersionDecider detectorNameVersionDecider;

    public DetectorEvaluationNameVersionDecider(DetectorNameVersionDecider detectorNameVersionDecider) {
        this.detectorNameVersionDecider = detectorNameVersionDecider;
    }

    public Optional<NameVersion> decideSuggestion(final List<DetectorEvaluation> detectorEvaluations, String projectDetector) {
        DetectorType preferredDetectorType = null;
        if (StringUtils.isNotBlank(projectDetector)) {
            final String projectDetectorFixed = projectDetector.toUpperCase();
            if (!DetectorType.POSSIBLE_NAMES.contains(projectDetectorFixed)) {
                logger.info("A valid preferred bom tool type was not provided, deciding project name automatically.");
            } else {
                preferredDetectorType = DetectorType.valueOf(projectDetectorFixed);
            }
        }

        final List<DetectorProjectInfo> allDetectorProjectInfo = transformIntoProjectInfo(detectorEvaluations);
        return detectorNameVersionDecider.decideProjectNameVersion(allDetectorProjectInfo, preferredDetectorType);
    }

    private List<DetectorProjectInfo> transformIntoProjectInfo(final List<DetectorEvaluation> detectorEvaluations) {
        return detectorEvaluations.stream()
                   .filter(it -> it.wasExtractionSuccessful())
                   .filter(it -> it.getExtraction().projectName != null)
                   .map(it -> {
                       final NameVersion nameVersion = new NameVersion(it.getExtraction().projectName, it.getExtraction().projectVersion);
                       final DetectorProjectInfo possibility = new DetectorProjectInfo(it.getDetector().getDetectorType(), it.getEnvironment().getDepth(), nameVersion);
                       return possibility;
                   })
                   .collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/synopsys/integration/detect/kotlin/nameversion/DetectorEvaluationNameVersionDecider.kt
package com.synopsys.integration.detect.kotlin.nameversion

import com.synopsys.integration.detector.base.DetectorEvaluation
import com.synopsys.integration.detector.base.DetectorType
import com.synopsys.integration.util.NameVersion
import org.codehaus.plexus.util.StringUtils
import org.slf4j.LoggerFactory
import java.util.*

class DetectorEvaluationNameVersionDecider(private val detectorNameVersionDecider: DetectorNameVersionDecider) {
    private val logger = LoggerFactory.getLogger(DetectorEvaluationNameVersionDecider::class.java)

    fun decideSuggestion(detectorEvaluations: List<DetectorEvaluation>, projectDetector: String?): Optional<NameVersion> {
        val detectorProjectInfo = detectorEvaluations
                .mapNotNull { toProjectInfo(it) }

        val detectorType = preferredDetectorTypeFromString(projectDetector)

        return detectorNameVersionDecider.decideProjectNameVersion(detectorProjectInfo, detectorType)
    }

    private fun preferredDetectorTypeFromString(detectorType: String?): DetectorType? {
        if (detectorType != null && detectorType.isNotBlank()) {
            val castDetectorType = detectorType.toUpperCase().takeIf { DetectorType.POSSIBLE_NAMES.contains(it) }?.let { DetectorType.valueOf(it) }
            if (castDetectorType != null) {
                logger.info("A valid preferred detector type was not provided, deciding project name automatically.")
            }
            return castDetectorType
        }
        return null
    }

    private fun toProjectInfo(detectorEvaluation: DetectorEvaluation): DetectorProjectInfo? {
        var projectName: String? = null
        var projectVersion: String? = null

        if (detectorEvaluation.wasDiscoverySuccessful()) {
            projectName = detectorEvaluation.discovery.projectName
            projectVersion = detectorEvaluation.discovery.projectVersion
        } else if (detectorEvaluation.wasExtractionSuccessful()) {
            projectName = detectorEvaluation.extraction.projectName
            projectVersion = detectorEvaluation.extraction.projectVersion
        }

        if (projectName != null) {
            if (StringUtils.isNotBlank(projectName)) {
                val nameVersion = NameVersion(projectName, projectVersion)
                return DetectorProjectInfo(detectorEvaluation.detectorRule.detectorType, detectorEvaluation.searchEnvironment.depth, nameVersion)
            }
        }

        return null
    }
}
